### ISSUE TITLE
Metanorma HTML output should be (A) valid and (B) accessible

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/isodoc#103]
Using the W3C validator (https://validator.w3.org/) , the Metanorma HTML output (https://standards.calconnect.org/csd/cc-10005.html) is shown to have 50 errors and 32 warnings.

Somehow the document is given as `xhtml`, not `html`. The HTML should certainly be HTML5 compliant.

<img width="754" alt="Screen Shot 2019-05-29 at 4 24 20 PM" src="https://user-images.githubusercontent.com/11865/58541500-37ee9580-822e-11e9-9d9f-1612d34b665d.png">

Validating as HTML5:

<img width="759" alt="Screen Shot 2019-05-29 at 4 24 39 PM" src="https://user-images.githubusercontent.com/11865/58541517-4341c100-822e-11e9-85c6-7e5f229aacf1.png">

1. This is a Metanorma-wide problem. How do we fix this throughout?

2. I know @pierrefortier is good at hand-crafting HTML -- @opoudjis is there a way to let @pierrefortier fix up these issues?

**Comment by ronaldtse** [comment]
@strogonoff to start with, the following flavors have these formats:

* CalConnect, OGC: HTML (normative/authoritative), PDF (informative), Word (informative)
* ISO, ITU, NIST, M3AAWG, UN, GB: Word (normative), HTML (informative), PDF (informative)
* IETF: XML (normative), HTML (informative)

**Comment by strogonoff** [comment]
@ronaldtse @opoudjis if you have this information at hand, could you point to the base code that generates the bulk of resulting HTML? The actual tag strings

I.e., the logic you assume will be changed as HTML validation issues are worked on.

I assume it’s something in https://github.com/metanorma/isodoc/tree/master/lib/isodoc?

**Comment by opoudjis** [comment]
The HTML is in isodoc/lib/isodoc/*, and in metanorma-*/lib/isodoc/base_converter.rb and metanorma-*/lib/isodoc/html_converter.rb

Given how monstrously long testing already takes to run, and given how formatting is tweaked all the time, I am even more hostile than usual to any pixel-perfect testing. Word HTML, as you'll already know from the documentation, is very different from HTML 5, and is idiosyncratic.

This is a requirement that requires reengineering of a very deep stack of HTML generation via Nokogiri. Given the effort vs impact tradeoff of it, I am not prioritising it.

**Comment by strogonoff** [comment]
I see, thanks! As to samples, I will use the ones we reference in authoring docs, but if you have comprehensive example projects specifically for layout compliance testing purposes, let me know

**Comment by strogonoff** [comment]
> Given the effort vs impact tradeoff of it, I am not prioritising it.

I’m currently acting under the assumption that I’m the one to take this (@ronaldtse)

> Given how monstrously long testing already takes to run, and given how formatting is tweaked all the time, I am even more hostile than usual to any pixel-perfect testing. Word HTML, as you'll already know from the documentation, is very different from HTML 5, and is idiosyncratic, and will never be made to conform to XHTML.
> 
> This is a requirement that requires reengineering of a very deep stack of HTML generation via Nokogiri.

Yeah, it’s hard to promise pixel-perfect consistency everywhere after HTML is validated. Right here in comments I’m trying to understand how strict are SDO requirements to visual layout, _so that Metanorma still delivers on the basic promise of compliant documents while generating valid HTML_. If I don’t get any specifics, I’ll have to proceed as is and wait for complaints later (it’ll be in a branch of course).

**Comment by strogonoff** [comment]
Most of the validation issues come down to a small number of patterns (such as the DOCTYPE and use of certain DOM element attributes). There’s progress in this direction.

I’ve combined this issue with the accessibility problem. Making markup valid is actually not that hard (most difficulty is in unraveling where in Ruby it is generated). Making it accessible may be the actual challenge afterwards.

**Comment by opoudjis** [comment]
@strogonoff Please let me know what the patterns are, and I'll patch them. I'm going to start working on the accessibility. This is now high priority, because NIST are asking for it.

**Comment by opoudjis** [comment]
I have had to convert the HTML to XHTML internally in isodoc, because it makes generating and processing the HTML much easier; but I certainly wouldn't want it to be subject to that strict a validation. I'm going to see about removing that DOCTYPE...

**Comment by opoudjis** [comment]
I am not going to fix the errors to do with comments, because I am still waiting for https://github.com/metanorma/isodoc-style/issues/3

I see that HTML5 validation means I have to mark up footnotes differently than with rel=footnote;  that will likely end up as just a class.

Lots of align attributes to be replaced by CSS.

@strogonoff, I'll take this over. I'm doing HTML5 rather than XHTML, I'd rather avoid the strictness.

**Comment by opoudjis** [comment]
In stylesheets (pending https://github.com/metanorma/isodoc/issues/112), need to correct 

````
writing-mode:tb-rl;
````

to 

````
writing-mode:vertical-rl;
````

**Comment by opoudjis** [comment]
The remain errors flagged in the Rice document are:

* Errors with annotations, which currently just used Word HTML; I am not going to do anything about them, because that is waiting on https://github.com/metanorma/isodoc-style/issues/3, to implement a true HTML solution.
* The non-compliant encoding of footnotes, which I'm not going to touch before https://github.com/metanorma/isodoc/issues/112 allows me to edit the associated CSS more efficiently.
* pre/p

**Comment by opoudjis** [comment]
`pre/p` occurs with annotations of callouts. Replaced paragraphs in annotations with the pure text content of annotations (stripping all formatting). Delimiting annotations from each other with `<br/>`, delimiting first annotation from sourcecode with `<br/><br/>`.

**Comment by opoudjis** [comment]
Will use https://github.com/pa11y/pa11y for accessibility testing.

**Comment by opoudjis** [comment]
Two remaining WCAGAA issues on the rice document reported by pa11y.

* Insufficient colour contrast for table rows. I consider a global review of CSS colours to be out of scope of this revision.
* The relationship between td elements and their associated th elements is not defined. Use either the scope attribute on th elements, or the headers attribute on td elements.

**Comment by opoudjis** [comment]
This means that any rowspan the elements need to be flagged with scope=col: https://www.w3.org/TR/WCAG20-TECHS/H63.html. That's robotic application of that criterion, but ok...

**Comment by opoudjis** [comment]
On hold pending #112: remaining fix is for footnotes.

**Comment by opoudjis** [comment]
Fixes are almost complete: removing ticket from on hold.

**Comment by opoudjis** [comment]
Removed epub namespace from HTML, last barrier to valid HTML 5.
